### PR TITLE
Cid image

### PR DESCRIPTION
**Pull request by burzum**
See #8877

### CODE REVIEW

File: src/View/Helper/HtmlHelper.php [hunk 1]
```diff
@@ -776,6 +776,21 @@ protected function _prepareCrumbs($startText, $escape = true)
     }
 
     /**
+     * Check if a given image path is a CID.
+     *
+     * @param $path array|string $path
```
Comment (dereuromark) [inline src/View/Helper/HtmlHelper.php]: order should be `type $var`.

**Comment by markstory** [comment]
You have [a few PHPCS errors](https://travis-ci.org/cakephp/cakephp/jobs/133364674#L647) but otherwise this looks good.

**Comment by ADmad** [comment]
Given how seldom this would be used to generate tags for embedded images, I don't like this overhead for all `image()` calls. Personally I don't see any benefit using the helper to generated such tags instead of straight up HTML.

**Comment by dereuromark** [comment]
I personally also always use a helper and afterRender callback to parse the images into CIDs and shove them into the Email class this way after the whole layout rendering is complete in the view layer.
Same I do with CSS to Inline CSS as described [here](https://github.com/dereuromark/cakephp-tools/wiki/Sending-templated-emails-with-CSS-and-inline-images),

Don't quite see the need for this either, but maybe I am missing something here.

**Comment by ndm2** [comment]
Besides `cid`, there are quite a few schemes that do not use `//`, so maybe this could incorportaed as general change to URL handling so that the helpers handle `abc:` schemes the same way as `abc://` ones, ie as absolute URLs that do not need any further processing?

**Comment by thinkingmedia** [comment]
The problem here is that each scheme has it's own special formats.

```
// Uniform Resource Identifier
urn:issn:1535-3613
// email
mailto:somebody@someplace.com
// wiki
wikipedia:Biome
// xri
xri://broadview.library.example.com/(urn:isbn:0-395-36341-1)
// posix
hostname:/directorypath/resource
// smb
smb://hostname/directorypath/resource
```

URLs in Cake are relative to the Router and nothing more. I think the issue here is that a string value was modified for a URL. I thought the Html helper wouldn't do that? Routing should only be done if an array is passed as the URL.

**Comment by ADmad** [comment]
> I think the issue here is that a string value was modified for a URL

String values also need to be modified to convert param `foo.jpg` to path `/img/foo.jpg`.

**Comment by ADmad** [comment]
For other special type of image path one can just usage `tag()` method like:

```
$this->Html->tag('img', ['src' => 'cid:cakephp_logo', ...]);
```

**Comment by markstory** [comment]
@ADmad That feels awkward to me

Is `cid:` the kind of prefix that could [live here with its friends](https://github.com/cakephp/cakephp/blob/master/src/Routing/Router.php#L614-L622)?

**Comment by ndm2** [comment]
@markstory Generally having the router recognize `cid` would be good I guess, and at that point I'd like to suggest a generic solution of recognizing `abc:` schemes again, since there are quite a few more schemes that would be nice to have.

As for the image issue, how about having the `UrlHelper::assetUrl()` bail out early for `abc:` as it does already for arrays or paths that contain `://`?

https://github.com/cakephp/cakephp/blob/3.2.10/src/View/Helper/UrlHelper.php#L123-L125
https://github.com/cakephp/cakephp/blob/3.2.10/src/View/Helper/UrlHelper.php#L138-L140

**Comment by markstory** [comment]
@ndm2 Yes that is another good option too.

**Comment by ADmad** [comment]
Putting the check in UrlHelper::assetUrl() as @ndm2 suggest is a better option.

**Comment by burzum** [comment]
@markstory what would you like to do with this one now? See the comments of @ADmad and @ndm2

**Comment by dereuromark** [comment]
I agree, if anything this should probably go to the Url assetUrl() method.

**Comment by markstory** [comment]
@burzum Handling it in UrlHelper would be nice.

**Comment by markstory** [comment]
Moving to #8966

**Pull request closed without merging**